realm-jni/build.gradle

```diff
@@ -89,18 +104,75 @@ task checkProperties(group: 'check', description: 'Check the user provided gradl
     }
 }
 
-// TODO: Split this task into smaller subtasks
+ext.coreArchiveFile = project.file("../core-android-${project.coreVersion}.tar.gz")
+ext.coreDir = new File(buildDir, "core-${project.coreVersion}")
+
+def coreDownloaded = false
+
 task downloadCore(group: 'build setup', description: 'Download the latest version of realm core', dependsOn: checkProperties) {
-    outputs.dir "../core-${project.coreVersion}"
+    outputs.upToDateWhen {
+        if (project.forceDownloadCore || !project.coreArchiveFile.exists()) {
+            return false
+        }
+        if (!isHashCheckingEnabled()) {
+            println 'skipped hash checking(empty \'coreSha256Hash\').'
+            return true
+        }
+
+        MessageDigest sha = MessageDigest.getInstance("SHA-256");
+        Formatter hexHash = new Formatter()
+        sha.digest(project.coreArchiveFile.bytes).each { b -> hexHash.format('%02x', b) }
+        return project.coreSha256Hash.equals(hexHash.toString())
+    }
+
     doLast {
+        if (project.forceDownloadCore) {
+            println 'forceDownloadCore is true.'
```

Comment:
Perhaps just "Downloading core..."? It shouldn't come as an surprise that forceDownload is set so no need to print it in the log IMO